### PR TITLE
[Sema] Only require a default implementation for SPI requirements in non-SPI protocols

### PR DESCRIPTION
**Pull request by xymus**
The check for SPI protocol requirements needing a default implementation should ignore requirements in a protocol that is SPI too. The protocol can’t be conformed to without knowing of the requirement.

rdar://65286171

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Review by beccadax — APPROVED** [review]
This change looks good. However, is there a runtime test verifying that test cases like `PublicProto` work correctly?

File: lib/Sema/TypeCheckAttr.cpp [hunk 1]
```diff
@@ -876,39 +876,41 @@ void AttributeChecker::visitSPIAccessControlAttr(SPIAccessControlAttr *attr) {
...
-          auto extension = dyn_cast<ExtensionDecl>(DC);
-
-          // The implementation must be defined in the same module in
-          // an unconstrained extension.
-          if (!extension ||
-              extension->getParentModule() != protocol->getParentModule() ||
-              extension->isConstrainedExtension())
-            return false;
+      if (!protocol->isSPI()) {
+        auto implementations = TypeChecker::lookupMember(
```
Comment (slavapestov) [inline lib/Sema/TypeCheckAttr.cpp:880]: This should reuse the existing logic for finding default implementations instead of rolling its own.

Comment (slavapestov) [reply lib/Sema/TypeCheckAttr.cpp:880]: More specifically I think these checks could be moved to typechecker::inferDefaultWitnesses().

Comment (xymus) [reply lib/Sema/TypeCheckAttr.cpp]: Thank you for pointing this out. I moved the check and it is much cleaner!

**Comment by xymus** [comment]
Moved the check to TypeCheckProtocol with pre-existing logic.

I'll have to spend more time on the runtime test. It works as expected with library evolution but it crashes in an ugly way without it when rebuilding from the public swiftinterface.

@swift-ci Please smoke test

**Pull request merged**